Include support for GQL custom scalar types

ahussain3 opened the pull request:
In the GraphQL schema, it is possible to specify (CustomScalarTypes)[https://www.apollographql.com/docs/graphql-tools/scalars.html].

In order to specify a Custom Scalar Type, you need to define an object which implements the following interface:
```
serialize()
parse_literal()
parse_value()
```
These methods allow GraphQL to translate between the serialized and deserialized versions of the custom scalar object.

However, when the GQL schema is passed around it includes only the names of the custom scalar types, and not the serialization/deserialization methods. After this PR is merged, you should be able to define your own classes on the client, and by giving them a parse_value() method you can use them as custom scalar types.

To configure the gql client to use your custom scalar types, you should do:
```
from gql import gql, Client
custom_types = {
    'DateTime': MyDateTimeClass
}

client = Client(schema=schema, transport=transport, custom_types=custom_types)
response = client.execute(query)
```

Commit:
Include support for GQL custom scalar types

Move the GQLResponseParser into the gql library.

Each time a GQL query is made by the client, the response is passed through
ResponseParser.parse() where it is processed.

Configure your own custom scalars as follows:
```
    custom_scalars = {
        'SomeCustomScalar': <ScalarClass>
    }

    gql_client = GQLClient(transport=gql_transport,
                           schema=schema,
                           custom_scalars=custom_scalars)

    gql_client.execute(...)
```
<ScalarClass> should have a .parse_value(value) function

There are a few anti-patterns I have had to include in order to support
some new functionality that we require:
- client.execute now accepts `url` and `headers` as arguments, since we
  need to be able to set these on a per request basis. Previously they were
  supplied by the transport (which gets set only once at initialization
  time).
- As a consequence, the url supplied in to the transport goes unused if a
  url is passed in to `execute()`. It is a required field so I have to pass
  in a string, but it's not the best.

lincolnq commented on gql/response_parser.py:
```diff
@@ -0,0 +1,114 @@
+from graphql.type.definition import GraphQLObjectType, GraphQLField, GraphQLScalarType
+
+
+class ResponseParser(object):
```
needs a different name - something involving type adapters or whatever

lincolnq commented on gql/transport/requests.py:
```diff
@@ -20,7 +20,7 @@ def __init__(self, url, auth=None, use_json=False, timeout=None, **kwargs):
         self.use_json = use_json
 
-    def execute(self, document, variable_values=None, timeout=None):
+    def execute(self, document, variable_values=None, timeout=None, url=None, headers={}):
```
url should not be a param. headers is ok, but make it a separate pr

lincolnq commented on gql/transport/requests.py:
```diff
@@ -29,13 +29,13 @@ def execute(self, document, variable_values=None, timeout=None):
         data_key = 'json' if self.use_json else 'data'
         post_args = {
-            'headers': self.headers,
+            'headers': headers or self.headers,
```
merge the dicts (use argument to override self.headers)

lincolnq commented on gql/response_parser.py:
```diff
@@ -0,0 +1,114 @@
+
+class ResponseParser(object):
+    """The challenge is to substitute custom scalars in a GQL response with their
+    decoded counterparts.
```
This should be written more clearly: explain briefly what "custom scalars" are, and what "decoded" means in this context.

lincolnq commented on gql/response_parser.py:
```diff
@@ -0,0 +1,114 @@
+
+    Public Interface:
+    parse(): call parse with a GQL response to replace all instances of custom
+        scalar strings with their deserialized representation."""
```
remove "call parse with"; we know it's a function :) (also rename the function)

lincolnq commented on tests/test_response_parser.py:
```diff
@@ -0,0 +1,94 @@
+"""Tests for the GraphQL Response Parser.
+
+These tests are worthless until I have a schema I can work with.
```
check in the starwars schema (augmented if necessary with some custom type definitions) and use it for testing, rather than depending on Money or whatever.

lincolnq reviewed (changes requested):
good progress, we spoke in person

Commit:
Fix failing tests

The tests were copied over from a different repo and so were inconsistent
with the Star Wars schema we are using here. This PR should fix the tests
(although we are still using a remote schema rather than a local one, so
the tests run slowly becuase they need to make a http request each time).

Commit:
Rename ResponseParser to TypeAdaptor

The 'ResponseParser' name was confusing because it was too broad, and we
weren't really actually parsing the response. Hopefully this makes it
clearer that what we are actually doing it just adding support for custom
types.

Commit:
Clean up docstrings in `type_adaptor` file

Stick to the format of having a one line summary followed by more detailed
information.

Commit:
Bugfix: Empty headers should default to empty dict

Since we are now merging the headers specified in the transport with the
headers specified on a per request basis, we need to make sure the old
headers are of type `dict`, a `None` is not sufficient.

Commit:
Add new exceptions (GQLSyntaxError and GQLServerError)

GQLServerError maps to errors which are generated on the backend server.
I.e. the server responds with a <200 OK> but an "errors" field is included
within the response. These are often user facing errors and should be
handled as such.

GQLSyntaxError maps to cases where either the query or the schema is
improperly formatted. This usually indicates a mistake in the code.

benkuhn commented on gql/client.py:
```diff
@@ -4,6 +4,8 @@
 from graphql.validation import validate
 
 from .transport.local_schema import LocalSchemaTransport
+from .type_adaptor import TypeAdaptor
```
Most Python projects spell it `adapter`. For consistency it's probably best to spell it that way ourselves too.

benkuhn commented on gql/client.py:
```diff
@@ -17,7 +19,7 @@ def __init__(self, retries_count, last_exception):
 class Client(object):
     def __init__(self, schema=None, introspection=None, type_def=None, transport=None,
-                 fetch_schema_from_transport=False, retries=0):
+                 fetch_schema_from_transport=False, custom_scalars={}, retries=0):
```
[nitpick] If we want to get this merged upstream, it's best to put new arguments at the end to preserve backwards compatibility of the function signature.

benkuhn replied:
Also, `custom_scalars` makes it sound like the dict values are scalars, rather than adapter functions. Maybe `custom_types` or something instead?

benkuhn replied:
PS: Also worth adding a docstring here that specifies the type of `custom_scalars`. (Annoying but otherwise it will be very hard to figure out!)

benkuhn commented on gql/type_adaptor.py:
```diff
@@ -0,0 +1,123 @@
+                return self._get_scalar_type_name(node)
+
+            final_node = self._follow_type_chain(node)
+            return iterate(final_node.fields[lookup.pop(0)], lookup)
```
Instead of mutation, it would be more idiomatic to use slicing here (and not copy above): `return iterate(final_node.fields[lookup[0]], lookup[1:])`--the way it's currently written, it's not immediately clear whether the second argument includes the original first key or not (you have to think through when the `pop` happens).

ahussain3 replied:
good point, thanks Ben

benkuhn commented on gql/type_adaptor.py:
```diff
@@ -0,0 +1,123 @@
+        By default the root level is `schema.query`, if that fails, then we check
+        `schema.mutation`."""
+
+        def iterate(node: Any, lookup: List[str]):
```
Can you give this a better name than `iterate`?

benkuhn commented on gql/type_adaptor.py:
```diff
@@ -0,0 +1,123 @@
+            final_node = self._follow_type_chain(node)
+            return iterate(final_node.fields[lookup.pop(0)], lookup)
+
+        try:
```
This `try/except` is icky (and could theoretically cause bugs if you have a query and a mutation with the same name, although that would be stupid). What if instead this function took an argument indicating whether the root type should be `query` or `mutation`?

ahussain3 replied:
I think it's also a little awkward to pass in the root type because there is no natural way to determine whether it is `query` or `mutation` (and I think it provides for a cleaner interface if this function is responsible for determining that). I can refactor to make it so that we explicitly work out if it is `query` or `mutation` rather than relying on an exception to be thrown.

benkuhn commented on gql/type_adaptor.py:
```diff
@@ -0,0 +1,123 @@
+            except (KeyError, AttributeError):
+                return None
+
+    def _substitute(self, keys: List[str], value: Any) -> Any:
```
This name makes it sound like the method mutates something. Maybe `_get_decoded_scalar_type` or something instead?

benkuhn commented on gql/type_adaptor.py:
```diff
@@ -0,0 +1,123 @@
+        modified."""
+        def iterate(node: Any, keys: List[str] = []):
+            if isinstance(node, dict):
+                result = {}
```
You can use a dict comprehension: `return {_key: iterate(value, keys + [_key]) for key, value in node.items()}`

benkuhn commented on tests/test_type_adaptor.py:
```diff
@@ -0,0 +1,118 @@
+    def parse_value(self, value: str):
+        return value.upper();
+
+@pytest.fixture
```
This makes the tests slower, dependent on the network, more likely to flake (if the network fails), and liable to break if the response at this URL changes.

Unless the project already depends on this particular URL, I'd suggest storing this in a file somewhere locally and parsing the schema yourself (the same way waveussd does it) instead of instantiating a client. In fact, even if this project already uses the URL elsewhere, it would probably be a good code improvement to remove that dependency :)

Either way, you should use `pytest.fixture(scope='session')` to prevent the fixture from being re-fetched on every request.

ahussain3 replied:
Yes, local schema would be best.

The project already depends on this URL (the same pattern is used elsewhere to get the schema), and we found that because of CSRF tokens it was quite cumbersome to get the schema via the standard tools (I tried the graphql-cli, and a json-to-graphql converter). So I left it like this to avoid wasting more time.

benkuhn commented on gql/type_adaptor.py:
```diff
@@ -0,0 +1,123 @@
+    apply(): pass in a GQL response to replace all instances of custom
+        scalar strings with their deserialized representation."""
+
+    def __init__(self, schema: GraphQLSchema, custom_scalars: Dict[str, Any] = {}) -> None:
```
Unfortunately, this project is supposed to work with Python 2.7 which doesn't support syntactic type annotations. I think it's probably worth it since this is eminently mergeable and then we could stop maintaining a fork.

benkuhn commented on gql/transport/requests.py:
```diff
@@ -20,16 +20,19 @@ def __init__(self, url, auth=None, use_json=False, timeout=None, **kwargs):
         self.use_json = use_json
 
-    def execute(self, document, variable_values=None, timeout=None):
+    def execute(self, document, variable_values=None, timeout=None, headers={}):
```
You should remove the custom headers change from this PR since it's in #2 as well. You can create a local branch with both PR's merged if you want something to run against locally (I think that's the best way to accomplish that, at least).

ahussain3 replied:
Will drop that commit from this branch 👍

benkuhn commented on gql/type_adaptor.py:
```diff
@@ -0,0 +1,123 @@
+        If keys (e.g. ['film', 'release_date']) points to a scalar type, then
+        this function returns the name of that type. (e.g. 'DateTime')
+
+        If it is not a scalar type (e..g a GraphQLObject or list), then this
```
Isn't it misleading to say the type could be a list (because we would just look at the list's `of_type` to see whether it's a list of scalars or objects)?

ahussain3 replied:
hmm, good point

benkuhn commented on gql/type_adaptor.py:
```diff
@@ -0,0 +1,123 @@
+        self.schema = schema
+        self.custom_scalars = custom_scalars
+
+    def _follow_type_chain(self, node: Any) -> Any:
```
This and `_get_scalar_type_name` could be pulled out of the class methods since they don't use `self`.

ahussain3 replied:
Is this always considered preferable? To me the functions seem relatively private, and I wouldn't imagine them needing to be called from outside of the class so I'm happy to leave them inside.

benkuhn commented on gql/type_adaptor.py:
```diff
@@ -0,0 +1,123 @@
+            return self.custom_scalars[scalar_type].parse_value(value)
+        return value
+
+    def _traverse(self, response: Dict[str, Any], substitute: Callable) -> Dict[str, Any]:
```
The code would be easier to read if `substitute` was hardcoded rather than an argument (since it's only ever equal to `self._substitute`) and if the name were less generic than `_traverse` (`_translate_all_scalars`, perhaps?)

benkuhn replied:
Actually, then `apply` (which also has an excessively generic name) would be a simple wrapper and you could remove it + make this the public function

benkuhn commented on tests/test_type_adaptor.py:
```diff
@@ -0,0 +1,118 @@
+    result = type_adaptor.apply(response)
+
+    assert result == expected
+    expected['allFilms']['edges'][0]['node']['releaseDate'] = "some_datetime"
```
What's the purpose of these two lines given that they're not followed by any assertions? Are they some kind of typo?

ahussain3 replied:
Mistake - will fix

benkuhn commented on gql/type_adaptor.py:
```diff
@@ -0,0 +1,123 @@
+        """ schema: a graphQL schema in the GraphQLSchema format
+            custom_scalars: a Dict[str, Any],
+                where str is the name of the custom scalar type, and
+                      Any is a class which has a `parse_value()` function"""
```
What's the signature of `parse_value`? It may be worth creating an abstract class that we expect the values in `custom_scalars` to subclass.

ahussain3 replied:
At least in our case that would be a little awkward since (for example) we already have a fleshed out CurrencyAmount class. The class would inherit from multiple base classes and CurrencyAmount would be tightly coupled to the gql client, which seems like it is more trouble than it is worth.

I will update the comment to make it clear what the signature of `parse_value` is

Commit:
Address CR changes

Commit:
Remove type annotations from TypeAdapter for py2.7 compatibility

Previously we were using Python3.5+ type annotation syntax but this is not
backwards compatible so we are removing it.
